### PR TITLE
[docs] MTP support

### PR DESCRIPTION
**Pull request by stevhliu**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47301)

Adds docs for enabling MTP on supported models (see https://github.com/huggingface/transformers/pull/46229)

### CODE REVIEW

File: docs/source/en/assisted_decoding.md [hunk 1]
```diff
@@ -141,6 +141,26 @@ outputs = model.generate(**inputs, assistant_early_exit=4, do_sample=False, max_
 tokenizer.batch_decode(outputs, skip_special_tokens=True)
 ```
 
+## Multi-token prediction
+
+Multi-token prediction (MTP) drafts candidate tokens with extra prediction layers trained into the model itself, so it does not need a separate assistant model. The MTP layers predict several future tokens from a single position and share the main model's embeddings and output head. The main model verifies the drafts in one forward pass, which speeds up generation without loading a second model.
```
Comment (vasqu) [inline docs/source/en/assisted_decoding.md]: > so it does not need a separate assistant model

I think this is an implementation detail and not super clear. Some libraries like vLLM do make separate implementations iirc. But the core layers are definitely the same. I think one of the major drawbacks for now is that it works on "standard" models but not on models like qwen 3.5 which have linear attention hybrid structures - so the cache is not yet compatible with that

File: docs/source/en/assisted_decoding.md [hunk 2]
```diff
@@ -141,6 +141,26 @@ outputs = model.generate(**inputs, assistant_early_exit=4, do_sample=False, max_
...
+```py
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+tokenizer = AutoTokenizer.from_pretrained("zai-org/GLM-4.5-Air")
+model = AutoModelForCausalLM.from_pretrained("zai-org/GLM-4.5-Air", device_map="auto")
+inputs = tokenizer("Hugging Face is an open-source company", return_tensors="pt").to(model.device)
+
+outputs = model.generate(**inputs, use_mtp=True, do_sample=False, max_new_tokens=50)
+tokenizer.batch_decode(outputs, skip_special_tokens=True)
```
Comment (vasqu) [inline docs/source/en/assisted_decoding.md:160]: Not sure yet but did we add training support in the end @Cyrilvallez?

**Review by vasqu — COMMENTED** [review]
Just some initial comments from my side, I think it's best to wait for @Cyrilvallez as I wasn't checking in the end anymore so only going from my mental model at that time :D

File: docs/source/en/assisted_decoding.md [hunk 3]
```diff
@@ -141,6 +141,26 @@ outputs = model.generate(**inputs, assistant_early_exit=4, do_sample=False, max_
 tokenizer.batch_decode(outputs, skip_special_tokens=True)
 ```
 
+## Multi-token prediction
```
Comment (Cyrilvallez) [inline docs/source/en/assisted_decoding.md]: Would put the acronym in title, as some people may know it but not necessarily the full term?
```suggestion

## Multi-Token Prediction (MTP)
```

File: docs/source/en/assisted_decoding.md (hunk 1 above)
Comment (Cyrilvallez) [reply docs/source/en/assisted_decoding.md]: Yeah, the way we implement it is also a "separate model" in a sense, but it's def an implem detail, as the layers are technically additional layers of the main model. It DOES work with any number of MTP layers and sliding window/linear conv layers since the changes I made for Inkling 🤗

File: docs/source/en/assisted_decoding.md [hunk 4]
```diff
@@ -141,6 +141,26 @@ outputs = model.generate(**inputs, assistant_early_exit=4, do_sample=False, max_
...
+
+Multi-token prediction (MTP) drafts candidate tokens with extra prediction layers trained into the model itself, so it does not need a separate assistant model. The MTP layers predict several future tokens from a single position and share the main model's embeddings and output head. The main model verifies the drafts in one forward pass, which speeds up generation without loading a second model.
+
+MTP works only with checkpoints trained with MTP layers, such as [DeepSeek-V3](https://huggingface.co/deepseek-ai/DeepSeek-V3) and [GLM-4.5](https://huggingface.co/zai-org/GLM-4.5). These models have a `num_mtp_layers` config value and MTP weights. [`~GenerationMixin.generate`] raises an error if the checkpoint has no MTP layers.
+
+Pass `use_mtp=True` to [`~GenerationMixin.generate`]. Like speculative decoding, MTP supports greedy search and sampling but not batched inputs.
+
```
Comment (Cyrilvallez) [inline docs/source/en/assisted_decoding.md:151]: Indeed! In the future, I plan to also add a `num_mtp_layers` option to be able to decide how many layers we want to use in the model has more than 1 (e.g. Inkling has 7 or 8 but usually it's not recommended to use all of them) - but this is not done yet

File: docs/source/en/assisted_decoding.md (hunk 2 above)
Comment (Cyrilvallez) [reply docs/source/en/assisted_decoding.md:160]: Yes, it should compute the correct loss needed to train the additional mtp layers, however this requires instantiating the mtp model and using its forward "separately" of the main model

**Review by Cyrilvallez — APPROVED** [review]
Nice, thanks for adding the doc for this nice new feature! Just added a few comments!
